### ISSUE TITLE
Crash on ObservableCollection<TKCustomPin>.Clear()

### ISSUE THREAD

**Issue by kolchy** [TorbenK/TK.CustomMap#4]
Hi Torben,

Just upgraded to 1.1.0 - FANTASTIC work :) :+1: One bug I've found so far is this crash when I try to clean a collection:

```
2015-12-10 09:50:32.706 MyApp.iOS[18325:1185660] Warning: Unhandled exception: System.InvalidCastException: Specified cast is not valid.
  at TK.CustomMap.iOSUnified.TKCustomMapRenderer.OnCollectionChanged (System.Object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs e) [0x00130] in <filename unknown>:0 
  at System.Collections.ObjectModel.ObservableCollection`1[T].OnCollectionChanged (System.Collections.Specialized.NotifyCollectionChangedEventArgs e) [0x00012] in /Users/builder/data/lanes/2377/73229919/source/mono/external/referencesource/System/compmod/system/collections/objectmodel/observablecollection.cs:286 
  at System.Collections.ObjectModel.ObservableCollection`1[T].OnCollectionReset () [0x00000] in /Users/builder/data/lanes/2377/73229919/source/mono/external/referencesource/System/compmod/system/collections/objectmodel/observablecollection.cs:373 
  at System.Collections.ObjectModel.ObservableCollection`1[T].ClearItems () [0x00022] in /Users/builder/data/lanes/2377/73229919/source/mono/external/referencesource/System/compmod/system/collections/objectmodel/observablecollection.cs:187 
  at System.Collections.ObjectModel.Collection`1[T].Clear () [0x00017] in /Users/builder/data/lanes/2377/73229919/source/mono/external/referencesource/mscorlib/system/collections/objectmodel/collection.cs:87 
  at MyApp.RootViewModel.UpdateVenueAndPins (System.Collections.Generic.List`1 venues) [0x0001d] in /Users/kolchy/Projects/MyApp/MyApp/ViewModels/RootViewModel.cs:76 
  at MyApp.RootViewModel.FindVenuesInBounds (Xamarin.Forms.Maps.MapSpan region) [0x001cd] in /Users/kolchy/Projects/MyApp/MyApp/ViewModels/RootViewModel.cs:251 
  at MyApp.MapView.CalculateBoundingCoordinates (Xamarin.Forms.Maps.MapSpan region) [0x00008] in /Users/kolchy/Projects/MyApp/MyApp/Views/MapView.cs:55 
  at MyApp.MapView+<Init>c__async0+<Init>c__AnonStorey7+<Init>c__async8.MoveNext () [0x00033] in /Users/kolchy/Projects/MyApp/MyApp/Views/MapView.cs:133 
```

The code from my viewmodel is:

```
if (Pins != null)
 Pins.Clear ();
```

Cheers,
Kolchy

**Comment by kolchy** [comment]
Had a quick look, and it could be trying to cast my custom pin to a TKCustomMapPin? I override the default TKCustomMapPin like so:

```
namespace MyApp
{
    public class CustomPin : TK.CustomMap.TKCustomMapPin
    {
        public Venue Venue { get; set; }
    }
}
```

as I need access to the venue ppl click on?

Cheers,
Kolchy

**Comment by kolchy** [comment]
One more observation, I found this worked :)

```
if (Pins != null) {
    var pins = Pins.ToList();
    foreach (var p in pins) {
        Pins.Remove (p);
    }
}
```

**Comment by TorbenK** [comment]
I just checked and it seems to work for me, at least on Android. I extended the sample in the development branch, you may want to have a look.

**Comment by yumshinetech** [comment]
Same issue here, It happens on real IOS device not simulator.

**Comment by yumshinetech** [comment]
```
private ObservableCollection<TKCustomMapPin> _mapPins = new ObservableCollection<TKCustomMapPin>();

        public ObservableCollection<TKCustomMapPin> MapPins
        {
            get { return _mapPins; }
            set
            {
                _mapPins = value;
                OnPropertyChanged(MapPinsPropertyName);
            }
        }
```

Invoke bellow code on somewhere

```
MapPin.Clear()
```

Throw an exception(Specified cast is not valid) when i test on ios device, but android device works fine.

Change MapPin.Clear() to following code:

```
foreach (var p in MapPin.ToList()) {
        MapPin.Remove (p);
    }
```

Throw an exception(Object not set to an instance of an object) when i test on android device, but ios device works fine.

**Comment by TorbenK** [comment]
Thank you for reporting. I will check and probably come up with a hotfix if possible.

**Comment by yumshinetech** [comment]
Hi TK,
I made some changes and resolved this issue,

Change:

```
else if (e.Action == NotifyCollectionChangedAction.Reset)
            {
                foreach (TKCustomMapAnnotation annotation in this.Map.Annotations)
                {
                    annotation.CustomPin.PropertyChanged -= OnPinPropertyChanged;
                }
                this._firstUpdate = true;
                this.UpdatePins();
            }
```

To:

```
else if (e.Action == NotifyCollectionChangedAction.Reset)
            {
                foreach (TKCustomMapAnnotation annotation in this.Map.Annotations.OfType<TKCustomMapAnnotation>())
                {
                    annotation.CustomPin.PropertyChanged -= OnPinPropertyChanged;
                }
                this._firstUpdate = true;
                this.UpdatePins();
            }
```

I am not sure that it is correctly changes.
Please take a look, I hope this can help you.

**Comment by TorbenK** [comment]
Hey, 

thanks for taking care of the issue. :+1: That is indeed the problem as it tries to cast annotations(like the current user position) to my internal annotation class, which obviously crashes. A major mistake on my part. I already pointed out the problem for Android as well. I will create a hotfix the next days.

Feel free to create a pull request for this and I will add you as contributor.

**Comment by yumshinetech** [comment]
I have created a new pull request and thanks for supporting.

**Comment by TorbenK** [comment]
Fixed in 1.2.1
